pyro/distributions/torch.py

```diff
@@ -3,197 +3,304 @@
 
 
 class Beta(torch.distributions.Beta, TorchDistributionMixin):
     def expand(self, batch_shape):
-        validate_args = self.__dict__.get('validate_args')
-        concentration1 = self.concentration1.expand(batch_shape)
-        concentration0 = self.concentration0.expand(batch_shape)
-        return Beta(concentration1, concentration0, validate_args=validate_args)
+        try:
+            sample_shape = self._sample_shape(batch_shape)
+            return self.expand_by(sample_shape)
+        except ShapeMismatchError:
+            validate_args = self.__dict__.get('validate_args')
+            concentration1 = self.concentration1.expand(batch_shape)
+            concentration0 = self.concentration0.expand(batch_shape)
+            beta = Beta(concentration1, concentration0, validate_args=validate_args)
+            beta.has_rsample = getattr(self, 'has_rsample')
+            return beta
 
 
 class Categorical(torch.distributions.Categorical, TorchDistributionMixin):
     def expand(self, batch_shape):
-        batch_shape = torch.Size(batch_shape)
-        validate_args = self.__dict__.get('validate_args')
-        if 'probs' in self.__dict__:
-            probs = self.probs.expand(batch_shape + self.probs.shape[-1:])
-            return Categorical(probs=probs, validate_args=validate_args)
-        else:
-            logits = self.logits.expand(batch_shape + self.logits.shape[-1:])
-            return Categorical(logits=logits, validate_args=validate_args)
+        try:
+            sample_shape = self._sample_shape(batch_shape)
+            return self.expand_by(sample_shape)
+        except ShapeMismatchError:
+            batch_shape = torch.Size(batch_shape)
+            validate_args = self.__dict__.get('validate_args')
+            if 'probs' in self.__dict__:
+                probs = self.probs.expand(batch_shape + self.probs.shape[-1:])
+                return Categorical(probs=probs, validate_args=validate_args)
+            else:
+                logits = self.logits.expand(batch_shape + self.logits.shape[-1:])
+                return Categorical(logits=logits, validate_args=validate_args)
 
 
 class Cauchy(torch.distributions.Cauchy, TorchDistributionMixin):
     def expand(self, batch_shape):
-        validate_args = self.__dict__.get('validate_args')
-        loc = self.loc.expand(batch_shape)
-        scale = self.scale.expand(batch_shape)
-        return Cauchy(loc, scale, validate_args=validate_args)
+        try:
+            sample_shape = self._sample_shape(batch_shape)
+            return self.expand_by(sample_shape)
+        except ShapeMismatchError:
+            validate_args = self.__dict__.get('validate_args')
+            loc = self.loc.expand(batch_shape)
+            scale = self.scale.expand(batch_shape)
+            cauchy = Cauchy(loc, scale, validate_args=validate_args)
+            cauchy.has_rsample = getattr(self, 'has_rsample')
```

Comment:
I suspect this is an incorrect fix for the `.has_rsample` issue, and that what we really want is
```py
return type(self)(loc, scale, validate_args=validate_args)
```
This way individual methods can be correctly overridden, but we must assume consistency of `__init__` args. WDYT?

Reply:
+1, I overlooked that the fakes implementation is just subclassing off of the distribution classes. This should be the correct fix. Will change!